presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -462,116 +366,265 @@ else if (!result.getBlocked().isDone()) {
+                        RemoteTask::getTaskId));
+        queryStateMachine.updateOutputLocations(bufferLocations, noMoreExchangeLocations);
     }
 
-    private StageExecutionAndScheduler getStageExecutionInfo(PlanFragmentId planFragmentId)
+    private static URI getBufferLocation(RemoteTask remoteTask, OutputBufferId rootBufferId)
     {
-        return stageExecutions.get(getStageId(planFragmentId));
+        URI location = remoteTask.getTaskStatus().getSelf();
+        return uriBuilderFrom(location).appendPath("results").appendPath(rootBufferId.toString()).build();
     }
 
-    private StageId getStageId(PlanFragmentId fragmentId)
+    private void addStateChangeListeners(SectionExecution sectionExecution)
+    {
+        for (StageExecutionAndScheduler stageExecutionAndScheduler : sectionExecution.getSectionStages()) {
+            SqlStageExecution stageExecution = stageExecutionAndScheduler.getStageExecution();
+            if (isRootFragment(stageExecution.getFragment())) {
+                stageExecution.addStateChangeListener(state -> {
+                    if (state == FINISHED) {
+                        queryStateMachine.transitionToFinishing();
+                    }
+                    else if (state == CANCELED) {
+                        // output stage was canceled
+                        queryStateMachine.transitionToCanceled();
+                    }
+                });
+            }
+            stageExecution.addStateChangeListener(state -> {
+                if (queryStateMachine.isDone()) {
+                    return;
+                }
+
+                if (state == FAILED) {
+                    ExecutionFailureInfo failureInfo = stageExecution.getStageExecutionInfo().getFailureCause()
+                            .orElseThrow(() -> new VerifyException(format("stage execution failed, but the failure info is missing: %s", stageExecution.getStageExecutionId())));
+                    Exception failureException = failureInfo.toException();
+
+                    boolean isRootSection = isRootFragment(sectionExecution.getRootStage().getStageExecution().getFragment());
+                    // root section directly streams the results to the user, cannot be retried
+                    if (isRootSection) {
+                        queryStateMachine.transitionToFailed(failureException);
+                        return;
+                    }
+
+                    if (retriedSections.get() >= maxStageRetries) {
+                        queryStateMachine.transitionToFailed(failureException);
+                        return;
+                    }
+
+                    if (!RECOVERABLE_ERROR_CODES.contains(failureInfo.getErrorCode())) {
+                        queryStateMachine.transitionToFailed(failureException);
+                        return;
+                    }
+
+                    try {
+                        if (sectionExecution.abort()) {
```

Comment:
`abort` returning false means already aborted does not seem intuitive, and my intuition is that returning false means the abort call failed (I should have commented on the commit when it is introduced). Maybe we should have a `isAborted()` in `SectionExecution`?

Reply:
we want the isAborted check and the abort() call to be atomic so that we only increment the retries once.  This is the same as the StateMachine transitionToAborted, which returns false if the state was already a done state.